### PR TITLE
core: avoid temporary memory allocation when parsing base58 CryptoHash

### PR DESCRIPTION
**Pull request by mina86**
When parsing base58-encoded CryptoHash, we first allocate a temporary
buffer, parse the encoded string into it, copy data from the buffer
into CryptoHash object and free the buffer.

This gets even wores with JSON parsing since then we first allocate
to hold the JSON string, then decode performing another allocation
and only then copy data into the CryptoHash buffer.

This all can be easily avoided.  So do that.

As a bonus, we get DoS protection which previously needed dedicated
condition for free.

### CODE REVIEW

File: core/primitives-core/src/hash.rs [hunk 1]
```diff
@@ -47,6 +46,31 @@ impl CryptoHash {
...
+/// Result of decoding base58-encoded crypto hash.
+enum Decode58Result {
+    /// Decoding succeeded.
+    Ok(CryptoHash),
+    /// The decoded data has incorrect length; either too short or too long.
+    BadLength,
+    /// There have been other decoding errors; e.g. an invalid character in the
+    /// input buffer.
+    Err(bs58::decode::Error),
 }
```
Comment (matklad) [inline core/primitives-core/src/hash.rs:74]: Not a huge fan of XxxResult which isn't a `std::result::Result`. 

Looking at it more holistically, it seems it should be enough to define just `FromStr` and `Display`, without any `base58_impl`, as those two APIs are general enough to not force allocation, and it should be possible to define serialization and whatnot in terms of those, and it seems possible?

```rust
impl fmt::Debug for CryptoHash {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "`{self}`")
    }
}

impl fmt::Display for CryptoHash {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut buffer = [0u8; 45];
        let len = bs58::encode(self).into(&mut buffer[..]).unwrap();
        let value = std::str::from_utf8(&buffer[..len]).unwrap();
        f.write_str(value)
    }
}

#[derive(thiserror::Error, Debug)]
pub enum ParseCryptoHashError {
    #[error("{0}")]
    InvalidBase58(bs58::decode::Error),
    #[error("invalid length")]
    InvalidLen,
}

impl std::str::FromStr for CryptoHash {
    type Err = ParseCryptoHashError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut buf = [0u8; 32];
        let len = bs58::decode(s).into(&mut buf).map_err(ParseCryptoHashError::InvalidBase58)?;
        if len != buf.len() {
            return Err(ParseCryptoHashError::InvalidLen);
        }
        Ok(CryptoHash(buf))
    }
}

impl<'de> Deserialize<'de> for CryptoHash {
    fn deserialize<D>(deserializer: D) -> Result<Self, <D as Deserializer<'de>>::Error>
    where
        D: Deserializer<'de>,
    {
        deserializer.deserialize_str(FromStrVisitor::<CryptoHash>::new())
    }
}

impl Serialize for CryptoHash {
    fn serialize<S>(&self, serializer: S) -> Result<<S as Serializer>::Ok, <S as Serializer>::Error>
    where
        S: Serializer,
    {
        serializer.collect_str(self)
    }
}

struct FromStrVisitor<T> {
    ghost: PhantomData<T>,
}

impl<T> FromStrVisitor<T> {
    fn new() -> FromStrVisitor<T> {
        FromStrVisitor { ghost: PhantomData }
    }
}

impl<'de, T> serde::de::Visitor<'de> for FromStrVisitor<T>
where
    T: FromStr,
    T::Err: fmt::Display,
{
    type Value = T;

    fn visit_str<E: serde::de::Error>(self, s: &str) -> Result<Self::Value, E> {
        s.parse::<T>().map_err(E::custom)
    }
}
```

That `FromStrVisitor` could be reused for other cases where we do Display<->Serialize thing (`PublicKey`, maybe something else?). 


I don't know a super-nice to pack that logic, I think this would sadly require a macro somewhere. But we can write that macro, or pull https://docs.rs/serde_with/. 


TL;DR: mild preference for making `FromStr/Display` the source of truth here, no  strong opinions on making that generalizable.

File: core/primitives-core/src/hash.rs [hunk 2]
```diff
@@ -47,6 +46,31 @@ impl CryptoHash {
...
+        }
+    }
+}
+
+/// Result of decoding base58-encoded crypto hash.
+enum Decode58Result {
+    /// Decoding succeeded.
+    Ok(CryptoHash),
+    /// The decoded data has incorrect length; either too short or too long.
+    BadLength,
```
Comment (nikurt) [inline core/primitives-core/src/hash.rs:70]: `BadLength` is conceptually an error. Why does it need to be represented by a separate enum variant?

**Review by nikurt — APPROVED** [review]

File: core/primitives-core/src/hash.rs (hunk 1 above)
Comment (mina86) [reply core/primitives-core/src/hash.rs:74]: Honestly I prefer my approach.  To me it is more readable with just two internal methods that everything is defined in terms of.  Especially since using Display goes through dynamic dispatch so it’s non-zero cost change to define everything in terms of Display.

File: core/primitives-core/src/hash.rs (hunk 2 above)
Comment (mina86) [reply core/primitives-core/src/hash.rs:70]: Yes, if this was a public API I would have wrapped it into another enum or change things.  But for an internal API I found this to be most straightforward approach.

File: core/primitives-core/src/hash.rs (hunk 1 above)
Comment (matklad) [reply core/primitives-core/src/hash.rs:74]: Yeah, this also makes sense. Internal method also gives us an extra guarantee, that the whole thing is a single `&str`.
